src/dask_awkward/lib/utils.py

```diff
@@ -0,0 +1,135 @@
+        elif form.is_list:
+            impl_with_parent(form.content, form, column_path)
+        elif form.is_option:
+            impl_with_parent(form.content, form, column_path)
+        elif form.is_record:
+            for field in form.fields:
+                impl_with_parent(form.content(field), form, column_path + (field,))
+        elif form.is_unknown or form.is_numpy:
+            pass
+        else:
+            raise AssertionError(form)
+
+    impl(form, ())
+
+    return {
+        "form_key_to_parent_key": form_key_to_parent_key,
+        "form_key_to_buffer_keys": form_key_to_buffer_keys,
+        "form_key_to_path": form_key_to_path,
+    }
+
+
+T = TypeVar("T")
+
+
+def walk_parents(node: T, parentage: dict[T, T | None]) -> Iterator[T]:
+    while (node := parentage.get(node)) is not None:
+        yield node
+
+
+def buffer_keys_required_to_compute_shapes(
+    parse_buffer_key: Callable[[str], tuple[str, str]],
+    shape_buffers: Iterable[str],
+    form_key_to_parent_key: dict[str, str],
+    form_key_to_buffer_keys: dict[str, str],
+):
+    # Buffers needing known shapes must traverse all the way up the tree.
+    for buffer_key in shape_buffers:
+        form_key, attribute = parse_buffer_key(buffer_key)
+
+        # For impacted form keys above this node
+        for impacted_form_key in walk_parents(form_key, form_key_to_parent_key):
+            # Identify the associated buffers
+            for impacted_buffer_key in form_key_to_buffer_keys[impacted_form_key]:
+                _, other_attribute = parse_buffer_key(impacted_buffer_key)
+
+                # Would the omission of this key lead to unknown lengths?
+                if other_attribute in UNKNOWN_LENGTH_ATTRIBUTES:
+                    # If so, touch the buffers
+                    yield impacted_buffer_key
+
+
+def render_buffer_key(form: Form, form_key: str, attribute: str) -> str:
+    return f"{form_key}-{attribute}"
+
+
+def parse_buffer_key(buffer_key: str) -> tuple[str, str]:
+    return buffer_key.rsplit("-", maxsplit=1)
+
+
+def form_with_unique_keys(form: Form, key: str) -> Form:
+    def impl(form: Form, key: str):
+        # Set form key
+        form.form_key = key
+
+        # If the form is a record we need to loop over all fields in the
+        # record and set form that include the field name; this will keep
+        # recursing as well.
+        if form.is_record:
+            for field in form.fields:
+                impl(form.content(field), f"{key}.{field}")
+
+        elif form.is_union:
+            for i, entry in enumerate(form.contents):
+                impl(entry, f"{key}#{i}")
+
+        # NumPy like array is easy
+        elif form.is_numpy or form.is_unknown:
+            pass
+
+        # Anything else grab the content and keep recursing
+        else:
+            impl(form.content, f"{key}.content")
+
+    # Perform a "deep" copy without preserving references
+    form = ak.from_dict(ak.to_dict(form))
```

Comment:
this line isn't working for me with `awkward@main` installed; should it be:?
```suggestion
    form = ak.forms.from_dict(form.to_dict())
```